### PR TITLE
enable Z-Image on Gaudi

### PR DESCRIPTION
**Pull request by nc-BobLee**

### CODE REVIEW

File: examples/stable-diffusion/text_to_image_zimage.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        help="Path to pre-trained model",
+    )
+    parser.add_argument(
+        "--prompts",
+        type=str,
+        nargs="*",
+        default="An image of a squirrel in Picasso style",
+        help="The prompt or prompts to guide the image generation.",
+    )
+    args = parser.parse_args()
```
Comment (Wei-Lin-Intel) [inline examples/stable-diffusion/text_to_image_zimage.py:59]: Add `width`, `height`, `seed`, `scale` and `steps` options here

File: examples/stable-diffusion/text_to_image_zimage.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    seed = 5451
+    random.seed(seed)
+    np.random.seed(seed)
+    torch.manual_seed(seed)
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--model_name_or_path",
+        default="Kolors",
```
Comment (Wei-Lin-Intel) [inline examples/stable-diffusion/text_to_image_zimage.py]: Change default to `Tongyi-MAI/Z-Image-Turbo`

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_zimage.py [hunk 3]
```diff
@@ -0,0 +1,668 @@
...
+
+        # From [batch, seq_len] to [batch, 1, 1, seq_len] -> broadcast to [batch, heads, seq_len, seq_len]
+        if attention_mask is not None and attention_mask.ndim == 2:
+            attention_mask = attention_mask[:, None, None, :]
+
+        query, key, value = (x.permute(0, 2, 1, 3) for x in (query, key, value))
+        query = query.contiguous()
+        key = key.contiguous()
+        value = value.contiguous()
+        hidden_states = FusedSDPA.apply(query, key, value, attention_mask, 0.0, False, None, "fast")
```
Comment (Wei-Lin-Intel) [inline optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_zimage.py]: Add a flag here to manually switch the softmax mode between `fast` and `fp32`. The default value is `fast`.

**Review by Wei-Lin-Intel — COMMENTED** [review]
Do we use fp32 rope or bf16 rope here?

File: examples/stable-diffusion/text_to_image_zimage.py (hunk 2 above)
Comment (nc-BobLee) [reply examples/stable-diffusion/text_to_image_zimage.py]: fp32 rope to align with original implementation.

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_zimage.py [hunk 4]
```diff
@@ -0,0 +1,670 @@
...
+        # From [batch, seq_len] to [batch, 1, 1, seq_len] -> broadcast to [batch, heads, seq_len, seq_len]
+        if attention_mask is not None and attention_mask.ndim == 2:
+            attention_mask = attention_mask[:, None, None, :]
+
+        query, key, value = (x.permute(0, 2, 1, 3) for x in (query, key, value))
+        query = query.contiguous()
+        key = key.contiguous()
+        value = value.contiguous()
+
+        softmax_mode = os.getenv("FP32_SOFTMAX_VISION", "fast")
```
Comment (Wei-Lin-Intel) [inline optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_zimage.py]:
```suggestion
        softmax_mode = 'fp32' if os.environ.get(
            'FP32_SOFTMAX_VISION', 'false').lower() in ['true', '1'
                                                             ] else 'fast'
```

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_zimage.py [hunk 5]
```diff
@@ -0,0 +1,670 @@
...
+            [`~pipelines.z_image.ZImagePipelineOutput`] or `tuple`: [`~pipelines.z_image.ZImagePipelineOutput`] if
+            `return_dict` is True, otherwise a `tuple`. When returning a tuple, the first element is a list with the
+            generated images.
+        """
+        height = height or 1024
+        width = width or 1024
+
+        vae_scale = self.vae_scale_factor * 2
+        if height % vae_scale != 0:
+            raise ValueError(
```
Comment (Wei-Lin-Intel) [inline optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_zimage.py]: Change `raise ValueError` into `Warning`, and then automatically adjust the width/height to the multiple of `vae_scale`

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_zimage.py [hunk 6]
```diff
@@ -0,0 +1,670 @@
...
+            self,
+            scheduler,
+            vae,
+            text_encoder,
+            tokenizer,
+            transformer,
+        )
+        n_refiner_layers =len(self.transformer.noise_refiner)
+        for i in range(n_refiner_layers):
+            self.transformer.noise_refiner[i] = ht.hpu.wrap_in_hpu_graph(self.transformer.noise_refiner[i])
```
Comment (Wei-Lin-Intel) [inline optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_zimage.py]: If `use_hpu_graphs` is True then you can apply `ht.hpu.wrap_in_hpu_graph` here, otherwise OOM issue will occur when resolution is very large.

File: optimum/habana/diffusers/models/attention_processor.py [hunk 7]
```diff
@@ -116,7 +116,7 @@ def forward(
         key_len = key.size(-2)
 
         # In the case of cross-attn, use FusedSDPA.
-        if  (query_len * cp_size) != key_len:
+        if  (query_len * cp_size) != key_len and query.size(0) < 8192:
```
Comment (Wei-Lin-Intel) [inline optimum/habana/diffusers/models/attention_processor.py]: or query_len < 8192

**Review by Wei-Lin-Intel — APPROVED** [review]
